nnlorun.py

```diff
@@ -344,23 +345,51 @@ def copy_to_grid(local_file, grid_file, args, maxrange=MAX_COPY_TRIES):
 
 
 def grid_copy(infile, outfile, args, maxrange=MAX_COPY_TRIES):
+    timeout_secs = 10
+    timeout = datetime.timedelta(minutes=timeout_secs)
+
     print_flush("Copying {0} to {1}".format(infile, outfile))
     protoc = args.gfaldir.split(":")[0]
     for protocol in PROTOCOLS: # cycle through available protocols until one works.
         infile_tmp = infile.replace(protoc, protocol)
         outfile_tmp = outfile.replace(protoc, protocol)
         print_flush("Attempting Protocol {0}".format(protocol))
+        outfile_dir = os.path.dirname(outfile_tmp)
+        outfile_fn = os.path.basename(outfile_tmp)
+        lscmd = "{gfal_loc}gfal-ls {dir}".format(dir=outfile_dir, gfal_loc=args.gfal_location)
+        
         for i in range(maxrange): # try max 10 times for now ;)
+            p = subprocess.check_output(lscmd, shell=True, universal_newlines=True).splitlines()
+            wait_until = datetime.datetime.now() + timeout
+            while outfile_fn in p and datetime.datetime.now() < wait_until:
```

Comment:
We could compare the checksum instead of simple look up

Reply:
Have a look at the new version, if you think it's ok then I'll make the same changes to the other *run.py files tomorrow